src/fs/metadata_log/entries.cc

```diff
@@ -0,0 +1,550 @@
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#include "fs/metadata_log/entries.hh"
+#include "fs/unix_metadata.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+using std::string_view;
+
+namespace seastar::fs::metadata_log::entries {
+
+enum ondisk_type : uint8_t {
+    INVALID = 0,
+    CHECKPOINT,
+    NEXT_METADATA_CLUSTER,
+    CREATE_INODE,
+    DELETE_INODE,
+    SMALL_WRITE,
+    MEDIUM_WRITE,
+    LARGE_WRITE_WITHOUT_MTIME,
+    LARGE_WRITE,
+    TRUNCATE,
+    CREADE_DENTRY,
+    CREATE_INODE_AS_DENTRY,
+    DELETE_DENTRY,
+    DELETE_INODE_AND_DENTRY,
+};
+
+template<class T, std::enable_if_t<std::is_standard_layout_v<T>, int> = 0>
+static void read_val(string_view& buff, T& val) noexcept {
+    assert(buff.size() >= sizeof(val));
+    std::memcpy(&val, buff.data(), sizeof(val));
+    buff.remove_prefix(sizeof(val));
+}
+
+template<class T, std::enable_if_t<std::is_standard_layout_v<T>, int> = 0>
+static void write_val(void*& dest, const T& val) noexcept {
+    std::memcpy(dest, &val, sizeof(val));
+    dest = reinterpret_cast<uint8_t*>(dest) + sizeof(val);
+}
+
+template<class size_type>
+static disk_offset_t ondisk_bstr_size(size_t data_len) noexcept {
+    return sizeof(size_type) + data_len;
+}
+
+template<class size_type, class Dest>
+[[nodiscard]] static std::optional<read_error> read_bstr(string_view& buff, Dest& data) noexcept {
+    size_type len;
+    read_val(buff, len);
+    if (buff.size() < len) {
+        return TOO_SMALL;
+    }
+
+    try {
+        using CharT = std::remove_const_t<std::remove_reference_t<decltype(data[0])>>;
+        static_assert(sizeof(CharT) == sizeof(char));
+        data = Dest(reinterpret_cast<const CharT*>(buff.data()), len);
+    } catch (...) {
+        return NO_MEM;
+    }
+
+    buff.remove_prefix(len);
+    return std::nullopt;
+}
+
+template<class size_type, class Src>
+static void write_bstr(void*& dest, const Src& data) noexcept {
+    assert(data.size() <= std::numeric_limits<size_type>::max());
+    write_val(dest, static_cast<size_type>(data.size()));
+    if constexpr (std::is_same_v<Src, std::string>) {
+        std::memcpy(dest, data.data(), data.size());
+    } else {
+        std::memcpy(dest, data.get(), data.size());
+    }
+    dest = reinterpret_cast<uint8_t*>(dest) + data.size();
+}
+
+#define READ_CHECKS(ONDISK_TYPE, ondisk_size_args...)                         \
```

Comment:
This can be quite easily transformed to a function with varargs which returns optional<err> - let's do that. I know that we end up with some boilerplate `if (err = something()) {return *err; }`, but we, and especially Avi, try to avoid macros at (almost) all costs, since they make debugging hard.

Reply:
Am I allowed to use macro `TRY_VAL` from `val_or_err.hh` during elimination of this macro? :)

Reply:
In general, macros are only tolerated either in testing code, or perhaps as a way of generating config entries. In case of regular code paths, the rule of thumb is to use 0 macros.

Reply:
In my opinion, macros allow to ease pain in some situations when there is no syntactic sugar in the language for such cases. Nonetheless, I agree that use of them should be minimized and will respect conventions for it here.